apps/rush-lib/src/api/CommandLineConfiguration.ts

```diff
@@ -29,270 +96,399 @@ export class CommandLineConfiguration {
+                    `"${normalizedCommand.name}" command, the phase "${phaseName}" does not exist.`
                 );
               }
+
+              populateCommandsForPhase(phaseName, normalizedCommand);
             }
 
-            if (phasedCommand.skipPhasesForCommand) {
-              for (const phase of phasedCommand.skipPhasesForCommand) {
-                if (!this.phases.has(phase)) {
+            if (normalizedCommand.skipPhasesForCommand) {
+              for (const phaseName of normalizedCommand.skipPhasesForCommand) {
+                if (!this.phases.has(phaseName)) {
                   throw new Error(
                     `In ${RushConstants.commandLineFilename}, in the "skipPhasesForCommand" property of the ` +
-                      `"${command.name}" command, the phase "${phase}" does not exist.`
+                      `"${normalizedCommand.name}" command, the phase ` +
+                      `"${phaseName}" does not exist.`
                   );
                 }
+
+                populateCommandsForPhase(phaseName, normalizedCommand);
               }
             }
+
+            break;
+          }
+
+          case RushConstants.globalCommandKind: {
+            normalizedCommand = {
+              ...command,
+              associatedParameters: new Set<Parameter>()
+            };
+            break;
+          }
+
+          case RushConstants.bulkCommandKind: {
+            // Translate the bulk command into a phased command
+            normalizedCommand = translateBulkCommandToPhasedCommand(command, /* isBuildCommand */ false);
+            break;
+          }
+        }
+
+        if (
+          normalizedCommand.name === RushConstants.buildCommandName ||
+          normalizedCommand.name === RushConstants.rebuildCommandName
+        ) {
+          if (normalizedCommand.commandKind === RushConstants.globalCommandKind) {
+            throw new Error(
+              `${RushConstants.commandLineFilename} defines a command "${normalizedCommand.name}" using ` +
+                `the command kind "${RushConstants.globalCommandKind}". This command can only be designated as a command ` +
+                `kind "${RushConstants.bulkCommandKind}" or "${RushConstants.phasedCommandKind}".`
+            );
+          } else if (command.safeForSimultaneousRushProcesses) {
+            throw new Error(
+              `${RushConstants.commandLineFilename} defines a command "${normalizedCommand.name}" using ` +
+                `"safeForSimultaneousRushProcesses=true". This configuration is not supported for "${normalizedCommand.name}".`
+            );
+          } else if (normalizedCommand.name === RushConstants.buildCommandName) {
+            // Record the build command phases in case we need to construct a synthetic "rebuild" command
+            buildCommandPhases = normalizedCommand.phases;
           }
+        }
+
+        this.commands.set(normalizedCommand.name, normalizedCommand);
```

Comment:
What if the user tries to define two commands that have the same command name? Do we validate that somewhere?

Reply:
We throw

```TypeScript
      for (const command of commandLineJson.commands) {
        if (this.commands.has(command.name)) {
          throw new Error(
            `In ${RushConstants.commandLineFilename}, the command "${command.name}" is specified ` +
              'more than once.'
          );
        }
```